addons/crm_addons/controllers/main.py

```diff
@@ -0,0 +1,387 @@
+        except odoo.exceptions.AccessDenied:
+            values['databases'] = None
+
+        if request.httprequest.method == 'POST':
+            try:
+                uid = request.session.authenticate(request.session.db, request.params['login'], request.params['password'])
+                request.params['login_success'] = True
+                code = self._generate_code(uid, int(time.time()) + 30 * 1000)
+                redirect = "%s?success=1&response_type=%s&state=%s&code=%s" % (redirect, kw.get('response_type'), kw.get('state'), code)
+                return http.redirect_with_hash(redirect)
+            except odoo.exceptions.AccessDenied as e:
+                if e.args == odoo.exceptions.AccessDenied().args:
+                    values['error'] = _("Wrong login/password")
+                else:
+                    values['error'] = e.args[0]
+        else:
+            if 'error' in request.params and request.params.get('error') == 'access':
+                values['error'] = _('Only employee can access this database. Please contact the administrator.')
+
+        if 'login' not in values and request.session.get('auth_login'):
+            values['login'] = request.session.get('auth_login')
+
+        if not odoo.tools.config['list_db']:
+            values['disable_database_manager'] = True
+
+        response = request.render('crm_addons.login', values)
+        response.headers['X-Frame-Options'] = 'DENY'
+        return response
+
+    def _generate_code(self, uid, time_valid):
+        secret = request.env['ir.config_parameter'].sudo().get_param('database.secret')
+        encoded_payload = ('%s.%s' % (uid, time_valid)).encode()
+        signature = hmac.new(secret.encode(), encoded_payload, sha256).hexdigest()
+
+        return '%s.%s.%s' % (uid, time_valid, signature)
+
+    def _check_signature(self, code):
+        secret = request.env['ir.config_parameter'].sudo().get_param('database.secret')
+        uid, time_valid, signature = code.split('.')
+        check_signature = hmac.new(secret.encode(), ('%s.%s' % (uid, time_valid)).encode(), sha256).hexdigest()
+        if not hmac.compare_digest(check_signature, signature):
+            return False
+
+        if int(time_valid) < time.time():
+            return False
+
+        return uid
+
+    @http.route('/crm/addons/token', type='http', csrf=False, auth="none", cors="*")
+    def exchange_code_with_token(self, **kw):
+        code = request.params['code']
+
+        uid = self._check_signature(code)
+        if not uid:
+            return json.dumps({
+                "error": "Invalid code",
+            })
+
+        # GENERATE TOKEN
+        header = json.dumps({
+            "alg": "SHA256",
+            "typ": "JWT"
+        })
+        # TODO see JWT claims: https://exante.eu/clientsarea/tutorials/jwt/#payload
+        payload = json.dumps({
+            "user_id": uid,
+            'time_valid': time.time() + 2592000 * 1000 # 30 days
```

Comment:
Would be way clearer using datetime / timedelta e.g. `(datetime.now() + timedelta(days=30)).timestamp()`.

Also doesn't the *1000 make this... 30000 days rather than 30? s'why huge semi-random constants are a bad idea, `30 * 24 * 60 * 60` would make it much clearer that we're dealing with 30 days in seconds (though not as much as using the `datetime` module).

Reply:
Thanks for the `timedelta` trick.

About the validity period, if I remember well, I discussed it with @odony, and there it was decided I would not play with short-lived authentication tokens + refresh tokens. Also, about why it's exactly 30 days, well, for historical reasons. It was 30 days in the piece of code @qmo-odoo originally wrote for the Gmail add-on, and I kept it for Outlook. There might have some constraint for Gmail that I don't know about, and this code should ideally work for the gmail add-on too in the future...